src/SparkConduit.sol

```diff
@@ -0,0 +1,314 @@
+    struct Position {
+        uint256 shares;
+        uint256 pendingWithdrawals;
+    }
+
+    // Please note totalShares/totalPendingWithdrawals are in aToken "shares"
+    // instead of the underlying asset
+    struct AssetData {
+        bool enabled;
+        uint256 totalShares;
+        uint256 totalPendingWithdrawals;
+        mapping (bytes32 => Position) positions;
+    }
+
+    // -- Storage --
+
+    mapping(address => AssetData) private assets;
+
+    /// @inheritdoc ISparkConduit
+    address public roles;
+    /// @inheritdoc ISparkConduit
+    address public registry;
+    /// @inheritdoc ISparkConduit
+    uint256 public subsidySpread;
+
+    // -- Immutable/constant --
+
+    /// @inheritdoc ISparkConduit
+    IPool   public immutable pool;
+    /// @inheritdoc ISparkConduit
+    address public immutable pot;
+
+    uint256 private constant RAY = 10 ** 27;
+    uint256 private constant SECONDS_PER_YEAR = 365 days;
+
+    modifier auth() {
+        require(wards[msg.sender] == 1, "SparkConduit/not-authorized");
+        _;
+    }
+
+    modifier ilkAuth(bytes32 ilk) {
+        require(
+            RolesLike(roles).canCall(ilk, msg.sender, address(this), msg.sig),
+            "SparkConduit/ilk-not-authorized"
+        );
+        _;
+    }
+
+    constructor(IPool  _pool, address _pot) {
+        pool = _pool;
+        pot  = _pot;
+    }
+
+    /// @inheritdoc IAllocatorConduit
+    function deposit(bytes32 ilk, address asset, uint256 amount) external ilkAuth(ilk) {
+        require(assets[asset].enabled, "SparkConduit/asset-disabled");
+        require(
+            assets[asset].positions[ilk].pendingWithdrawals == 0,
+            "SparkConduit/no-deposit-with-pending-withdrawals"
+        );
+        require(amount <= maxDeposit(ilk, asset), "SparkConduit/max-deposit-exceeded");
+
+        address source = RegistryLike(registry).buffers(ilk);
+
+        require(
+            IERC20(asset).transferFrom(source, address(this), amount),
+            "SparkConduit/transfer-failed"
+        );
+
+        pool.supply(asset, amount, address(this), 0);
+
+        // Convert asset amount to shares
+        uint256 shares = amount.rayDiv(pool.getReserveData(asset).liquidityIndex);
+
+        assets[asset].positions[ilk].shares += shares;
+        assets[asset].totalShares           += shares;
+
+        emit Deposit(ilk, asset, source, amount);
+    }
+
+    /// @inheritdoc IAllocatorConduit
+    function withdraw(bytes32 ilk, address asset, uint256 maxAmount)
+        external ilkAuth(ilk) returns (uint256 amount)
+    {
+        uint256 liquidityAvailable
+            = IERC20(asset).balanceOf(pool.getReserveData(asset).aTokenAddress);
+
+        amount = liquidityAvailable < maxAmount ? liquidityAvailable : maxAmount;
+
+        // Normally you should update local state first for re-entrancy,
+        // but we need an update-to-date liquidity index for that
```

Comment:
Lets add a commnet about the requireement to update the liquidityIndex in order to get the shares that correspond to the `amount` is

Reply:
This is no longer a requirement. I'm going to move the withdraw to the end of the function actually.

Reply:
This has been re-arranged and simplified.